### PR TITLE
coap-gateway/http-gateway: allow user to onboard device via authoriza…

### PR DESCRIPTION
**Pull request by jkralik**
…tion code from UI

### CODE REVIEW

File: http-gateway/service/config.go [hunk 1]
```diff
@@ -114,36 +114,63 @@ func (c *ClientsConfig) Validate() error {
...
 }
 
-func (c *OAuthClientConfig) Validate() error {
-	if c.Domain == "" {
-		return fmt.Errorf("domain('%v')", c.Domain)
-	}
+func (c *BasicOAuthClient) Validate() error {
 	if c.ClientID == "" {
-		return fmt.Errorf("clientID('%v')", c.ClientID)
+		return fmt.Errorf("domain('%v')", c.ClientID)
```
Comment (Danielius1922) [inline http-gateway/service/config.go]: shouldn't this be clientID?

File: coap-gateway/service/refreshToken.go [hunk 2]
```diff
@@ -89,11 +89,21 @@ func refreshTokenPostHandler(req *mux.Message, client *Client) {
 		return
 	}
 
-	token, err := client.server.provider.Refresh(req.Context, refreshToken.RefreshToken)
+	var token *oauth2.Token
+	for _, provider := range client.server.providers {
+		token, err = provider.Refresh(req.Context, refreshToken.RefreshToken)
```
Comment (Danielius1922) [inline coap-gateway/service/refreshToken.go]: maybe a log if `err != nil` or is it expected behavior for only one provider to succeed?

File: http-gateway/service/config.go [hunk 3]
```diff
@@ -114,36 +114,63 @@ func (c *ClientsConfig) Validate() error {
...
+}
+
+type DeviceOAuthClient struct {
+	BasicOAuthClient `yaml:",inline"`
+	ProviderName     string `json:"providerName" yaml:"providerName"`
+}
+
+func (c *DeviceOAuthClient) Validate() error {
+	if c.ClientID == "" {
+		return fmt.Errorf("domain('%v')", c.ClientID)
```
Comment (Danielius1922) [inline http-gateway/service/config.go]: clientID

File: http-gateway/service/config.go [hunk 4]
```diff
@@ -114,36 +114,63 @@ func (c *ClientsConfig) Validate() error {
...
+	}
+	return nil
```
Comment (Danielius1922) [inline http-gateway/service/config.go:142]: `return BasicOAuthClient.Validate()`

File: bundle/run.sh [hunk 5]
```diff
@@ -591,13 +591,14 @@ cat /configs/coap-gateway.yaml | yq e "\
...
   .apis.coap.externalAddress = \"${FQDN}:${COAP_GATEWAY_UNSECURE_PORT}\" |
   .apis.coap.tls.enabled = false |
-  .apis.coap.authorization.authority = \"https://${OAUTH_ENDPOINT}\" |
-  .apis.coap.authorization.clientID = \"${DEVICE_OAUTH_CLIENT_ID}\" |
-  .apis.coap.authorization.clientSecret = \"${DEVICE_OAUTH_CLIENT_SECRET}\" |
-  .apis.coap.authorization.redirectURL = \"${DEVICE_OAUTH_REDIRECT_URL}\" |
-  .apis.coap.authorization.scopes = [ \"${DEVICE_OAUTH_SCOPES}\" ] |
-  .apis.coap.authorization.audience = \"${DEVICE_OAUTH_AUDIENCE}\" |
-  .apis.coap.authorization.http.tls.useSystemCAPool = true |
+  .apis.coap.authorization.providers[0].name = \"plgd\" |
```
Comment (Danielius1922) [inline bundle/run.sh]: There is an env variable named `DEVICE_PROVIDER` for this.

File: http-gateway/service/config.go (hunk 4 above)
Comment (jkralik) [reply http-gateway/service/config.go:142]: for device provider audience is optional, so this is why DeviceOAuthClient has owned the `Validate` function.

File: http-gateway/service/config.go (hunk 1 above)
Comment (jkralik) [reply http-gateway/service/config.go]: fixed

File: coap-gateway/service/refreshToken.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/service/refreshToken.go]: we don't know which provider is needed to use so we try all of them

File: http-gateway/service/config.go (hunk 3 above)
Comment (jkralik) [reply http-gateway/service/config.go]: fixed

File: bundle/run.sh (hunk 5 above)
Comment (jkralik) [reply bundle/run.sh]: fixed

**Review by Danielius1922 — APPROVED** [review]

**Review by ondrejtomcik — APPROVED** [review]

**Pull request merged**